### PR TITLE
Mapchecker

### PR DESCRIPTION
**Pull request by TsjipTsjip**
## About the PR

This is a map checker! @dvir001 @Cheackraze 

Essentially this allows us to automatically check shuttle/POI/... PR's for prototypes which are "suspicious" for that map, automatically as a github workflow.

Additionally, the following was the output on script creation, after some heavy cleanup:
```
[ERROR - 2024-01-01 21:05:28] Found 13 maps with illegal prototypes.
[ERROR - 2024-01-01 21:05:28] Map 'Resources/Prototypes/_NF/Shipyard/sprinter.yml' has 2 illegal prototypes.
[ERROR - 2024-01-01 21:05:28]  - WallPlastitanium
[ERROR - 2024-01-01 21:05:28]  - IntercomSecurity
[ERROR - 2024-01-01 21:05:28] Map 'Resources/Prototypes/_NF/Shipyard/courser.yml' has 1 illegal prototypes.
[ERROR - 2024-01-01 21:05:28]  - HighSecDoor
[ERROR - 2024-01-01 21:05:28] Map 'Resources/Prototypes/_NF/Shipyard/metastable.yml' has 1 illegal prototypes.
[ERROR - 2024-01-01 21:05:28]  - SignSec
[ERROR - 2024-01-01 21:05:28] Map 'Resources/Prototypes/_NF/Shipyard/dartx.yml' has 1 illegal prototypes.
[ERROR - 2024-01-01 21:05:28]  - HighSecDoor
[ERROR - 2024-01-01 21:05:28] Map 'Resources/Prototypes/_NF/Shipyard/decadedove.yml' has 1 illegal prototypes.
[ERROR - 2024-01-01 21:05:28]  - SignSec
[ERROR - 2024-01-01 21:05:28] Map 'Resources/Prototypes/_NF/Shipyard/gourd.yml' has 1 illegal prototypes.
[ERROR - 2024-01-01 21:05:28]  - HoloprojectorSecurity
[ERROR - 2024-01-01 21:05:28] Map 'Resources/Prototypes/_NF/Shipyard/bison.yml' has 4 illegal prototypes.
[ERROR - 2024-01-01 21:05:28]  - WallPlastitanium
[ERROR - 2024-01-01 21:05:28]  - WallPlastitaniumDiagonal
[ERROR - 2024-01-01 21:05:28]  - SignSec
[ERROR - 2024-01-01 21:05:28] Map 'Resources/Prototypes/_NF/Shipyard/esquire.yml' has 3 illegal prototypes.
[ERROR - 2024-01-01 21:05:28]  - ToyFigurineSecurity
[ERROR - 2024-01-01 21:05:28]  - AirlockSecurity
[ERROR - 2024-01-01 21:05:28]  - SignSec
[ERROR - 2024-01-01 21:05:28] Map 'Resources/Prototypes/_NF/Shipyard/praeda.yml' has 4 illegal prototypes.
[ERROR - 2024-01-01 21:05:28]  - ClothingEyesGlassesSecurity
[ERROR - 2024-01-01 21:05:28]  - SignSec
[ERROR - 2024-01-01 21:05:28] Map 'Resources/Prototypes/_NF/Shipyard/pheonix.yml' has 3 illegal prototypes.
[ERROR - 2024-01-01 21:05:28]  - DebugSubstation
[ERROR - 2024-01-01 21:05:28]  - WallPlastitanium
[ERROR - 2024-01-01 21:05:28] Map 'Resources/Prototypes/_NF/Shipyard/spectre.yml' has 1 illegal prototypes.
[ERROR - 2024-01-01 21:05:28]  - AirlockSecurity
[ERROR - 2024-01-01 21:05:28] Map 'Resources/Prototypes/_NF/Shipyard/bazaar.yml' has 1 illegal prototypes.
[ERROR - 2024-01-01 21:05:28]  - AirlockSecurity
[ERROR - 2024-01-01 21:05:28] Map 'Resources/Prototypes/_NF/Shipyard/anchor.yml' has 1 illegal prototypes.
[ERROR - 2024-01-01 21:05:28]  - AirlockSecurity
```

All of these have either been tagged as allowable for this shipyard category, or have been temporarily whitelisted (technical debt).

## Why / Balance

Automation of a menial task which can overlook things in the hands of man.

## Technical details

- Github workflow runs `.github/mapchecker/mapchecker.py`
- Paths for prototype collection and map collection are specified in arguments which have been set to sensible defaults.
- Actual things to blacklist have to be configured in `.github/mapchecker/config.py` due to their complex nature.

## Media

- [ ] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

No. I have not. For my sanity I will do this later.

## Breaking changes

Any other PR that modifies any of these ships will likely be merge conflicted by this PR.

**Changelog**

(Deliberately not properly formatting this for the bot, this is a joke.)
- remove: Random stuff that shouldn't have been on ships in the first place
- add: Automation, welcome to the future

### CODE REVIEW

**Comment by TsjipTsjip** [comment]
Proooobably should also @MagnusCrowe here.

**Review by dvir001 — APPROVED** [review]
Going to wait for others to also approve this, seem fine by me.

**Review by Cheackraze — CHANGES_REQUESTED** [review]
try to keep ship file updates to their own PRs, else you will end up in conflict hell.

File: .github/mapchecker/whitelist.yml [hunk 1]
```diff
@@ -0,0 +1,46 @@
+Frontier:
```
Comment (Cheackraze) [inline .github/mapchecker/whitelist.yml:1]: for some of these, we might want the whole POI to just be skipped, like the frontier outpost or the upcoming nfsd poi

**Comment by TsjipTsjip** [comment]
> try to keep ship file updates to their own PRs, else you will end up in conflict hell.

Can do, will add all current violations to technical debt list of the whitelist to prevent this PR from murdering the success state of other PR's.

File: .github/mapchecker/whitelist.yml (hunk 1 above)
Comment (TsjipTsjip) [reply .github/mapchecker/whitelist.yml:1]: Can do as well, but the intention for this is for these kinds of changes to stand out. It's not foolproof, something that was already whitelisted can be added infinitely.

If someone wants to add a restricted prototype to their ship, I want them to have to edit the whitelist file to make it stand out that they are requesting that exception from regular mapping rules. Adding this based on the check output is pretty easy, as it tells you exactly which map name needs to have exactly which prototype names addressed.

**Review by Cheackraze — APPROVED** [review]

**Pull request merged**